### DISCUSSION TITLE
Adding Wave spinner

### DISCUSSION THREAD

**Discussion by H999** [Textualize/rich#2142] [Ideas]
Hi, I just updated pip today and looked at the src saw it use spinner so pretty, and I was directed here by pip.
At the file rich\_spinners.py I wondered why you don't have the wave spinner?

Adding in rich\_spinners.py (with the original spirit of using standard library):
```py
...
# MIT License
import itertools
import random

wave_frames=[''.join(item) for item in itertools.product('▁▂▃▄▅▆▇█', repeat=4) if len(set(item)) == 4]
random.shuffle(wave_frames)

SPINNERS = {
    "wave": {
        "interval": 120,
        "frames": wave_frames,
    },
...
}
```

and test with this:
```py
from rich.progress import (
    BarColumn,
    Progress,
    ProgressColumn,
    SpinnerColumn,
    TextColumn,
    TimeRemainingColumn,
    DownloadColumn,
    FileSizeColumn,
    TimeElapsedColumn,
    TransferSpeedColumn,
)
import time


def test(style):
    total = 100
    step = 1
    columns = (
        TextColumn("test [progress.description]{task.description}"),
        TimeElapsedColumn(),
        BarColumn(),
        SpinnerColumn(spinner_name='wave', finished_text=f'[{style}]{{████}}', style=style),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        TextColumn("eta","[progress.percentage]"),
        TimeRemainingColumn(),
        DownloadColumn(),
        FileSizeColumn(),
        TransferSpeedColumn(),
    )
    progress = Progress(*columns, refresh_per_second=1000)
    task_id = progress.add_task(" ", total=total)
    with progress:
        for chunk in range(total):
            # yield chunk
            time.sleep(0.000000127)
            # if not chunk%1000:
            #     print(chunk)
            progress.update(task_id, advance=step)

test("green")
```

it look quite ok, like the music is playing
![image](https://user-images.githubusercontent.com/32168478/161231425-06534738-c8e7-43de-aafa-2e135d7bd0db.png)

I think it will be nicer when you have it in core lib, so what do you think if you apply it

Note:
        - When use itertools.product I think with 4 char is enough, larger is slower to process
        - For future the wave spinner should have the random.shuffle where it used not init 1 time in rich\_spinners.py

**Comment by willmcgugan** [comment]
Can I see a video?

**Reply by H999** [reply]
ok, here you are

https://user-images.githubusercontent.com/32168478/161288041-2986a4d5-0c83-4c18-b7c0-42d2365c090d.mp4

**Comment by willmcgugan** [comment]
Doesn't look that great TBH. Now if it was a sine wave, that might be more interesting.

**Reply by H999** [reply]
thanks, but if it is a sine wave I don't think it more interesting, because it will use the static frames don't have the random look like music play like this. How about the idea how do you think it good enough to have in main lib. If you want I will find the way to sort it to be sine wave

**Reply by H999** [reply]
you can try this with smooth sine wave
```py
SPINNERS = {
    "wave_sine": {
        "interval": 120,
        "frames": ["▁▂▃▄", "▂▃▄▅", "▃▄▅▆", "▄▅▆▇", "▅▆▇█", "▆▇█▇", "▇█▇▆", "█▇▆▅", "▇▆▅▄", "▆▅▄▃", "▅▄▃▂", "▄▃▂▁", "▃▂▁▂", "▂▁▂▃"],
    },
...
}
```

https://user-images.githubusercontent.com/32168478/161309017-29b48609-290c-43c5-a346-459ef109e149.mp4

**Reply by willmcgugan** [reply]
I like that. Feel free to submit a PR.

**Reply by H999** [reply]
=]]]] maybe I keep both of them, I will submit a PR later

**Comment by H999** [comment]
hello @willmcgugan, I submitted a PR [here](https://github.com/Textualize/rich/pull/2143).
After submitted, I found that src have spinner growVertical similar with wave have 1 size length but the frames not smooth much I think you should update the frames from "▁▃▄▅▆▇▆▅▄▃" to "▁▂▃▄▅▆▇█▇▆▅▄▃▂".
And I not sure where should I locate the def init the frames so I put it in rich\_spinners.py because it do only 1 mission is init the waves frames, if you think somewhere else is better you can relocate it

**Comment by FedeClaudi** [comment]
The sine wave one is very cool, I might steal it for [Term](https://github.com/FedeClaudi/Term.jl) 😬

**Reply by H999** [reply]
=]]]]] Really, it would be my honor for my code to take part in your prj, don't be shy when use it because I think anybody can do it soon or late if they know code. If I push a PR to here, it mean I want to contribute and accept the MIT license, so just feel free to use it (except rich team don't want - but I don't think this situation will have 🤣).
ps: oh! you use the Julia, it very nice, I was planning to learn it but I don't have time, if have something to say about that is very derailed from the track 😅

**Comment by H999** [comment]
But really TBH, you are don't like the music wave like [this](https://www.youtube.com/watch?v=-qtZ5oiJVqE). With me it awesome more  than the sine wave.
